레스토랑 상세 페이지 구현하기

solmatches opened the pull request:
안녕하세요 6주차 과제 제출합니다!

location.path 값에 따라 다른 컴포넌트를 렌더하는 부분까지 구현하였고
이후 react-router를 적용하려고 합니다 🙂 

강의에서처럼 router v5로 구현하는게 좋을까요 아니면 v6로 구현 해보는 것도 괜찮을까요?

solmatches committed:
웹팩 devServer historyApiFallback 설정 추가

solmatches committed:
react-router-dom 설치

solmatches committed:
HomePage 컴포넌트 테스트 작성

solmatches committed:
RetaurantPage 컴포넌트 테스트 작성

solmatches committed:
RestaurantsPage 구현하고 App에서 가져와 적용

solmatches committed:
location.pathname에 따라 보여줄 컴포넌트 분기

hannut91 commented:
최신 버전이 V6로 구현하는게 좋을 것 같습니다

hannut91 commented on webpack.config.js:
```diff
@@ -14,4 +14,9 @@ module.exports = {
   resolve: {
     extensions: ['.js', '.jsx'],
   },
+  devServer: {
+    historyApiFallback: {
+      index: 'index.html',
+    },
+  },
```
이 문서도 읽어 봅시다

https://github.com/bripkens/connect-history-api-fallback#introduction

hannut91 reviewed (commented):
수업 내용을 잘 익히신 것 같네요 ㅎㅎ 라이브러리를 사용하면서 어떤점이 변화는지 유심히 지켜보면서 과제를 진행해 보세요

solmatches committed:
App 테스트에 기존 리덕스 mock 제거하고 해피패스 구현

solmatches committed:
AboutPage 테스트 작성

solmatches committed:
About 페이지 해피패스 구현

solmatches committed:
NotFoundPage 테스트 작성

solmatches committed:
NotFoundPage 페이지 제목 추가로 해피패스 구현

solmatches committed:
App에 About, Restaurant Page 추가하고 pathname 구조분해 변경

- window.location.pathname 구조 분해시 쓰던 방법대로만 사용하여 버릇이 된 것 같아
강의에 나온대로 방법을 변경해봄

solmatches committed:
ReactRouter 적용

solmatches committed:
페이지 컴포넌트들을 `pages` 디렉토리로 이동

페이지 컴포넌트들만 모아놓고 관리하기 위함

solmatches committed:
App ReactRouter 적용으로 테스트 추가

solmatches committed:
HomePage 페이지 제목, 목차 렌더 테스트 작성

solmatches committed:
HomePage 컴포넌트에 페이지 목차 구현

- Link가 사용되어 테스트 에러 발생
  - 테스트코드에 테스트용 Router로 빠른 MemoryRouter 보강

solmatches committed:
경로에 따라 다른 페이지 노출 확인 App 테스트 작성

solmatches committed:
페이지마다 헤더가 보이도록 App 테스트 작성

solmatches committed:
App 컴포넌트에 헤더 구현 및 Outlet으로 페이지 전환 부분 적용

- App에서 전체 페이지 공통 컴포넌트와 Outlet만 관리하도록
Routes, Route를 index.js로 이동
- 공통 헤더 추가가 되었으므로 시맨틱한 구조를 위해 다른 페이지들의 h1 태그를 -> h2로 교체

solmatches committed:
공통 헤더 컴포넌트 클릭시 홈으로 이동 테스트 작성

solmatches committed:
Header 컴포넌트 추가 및 클릭시 홈으로 이동 구현

- Link 사용으로 인해 테스트 오류 발생
  - 테스트에 MemoryRouter 추가
- Header 컴포넌트에서 HomePage 렌더링까지 확인하는 것이 컴포넌트 역할보다 오버되어
window.location.path 값 직접 확인으로 변경

hannut91 commented on src/pages/HomePage.test.jsx:
```diff
@@ -0,0 +1,24 @@
+import { render } from '@testing-library/react';
+import { MemoryRouter } from 'react-router-dom';
+import HomePage from './HomePage';
```
여기 임포트도 정리해 줍시다

hannut91 commented on src/pages/AboutPage.test.jsx:
```diff
@@ -0,0 +1,8 @@
+test('AboutPage', () => {
+  const { getByRole } = render(<AboutPage />);
+
+  expect(getByRole('heading')).toHaveTextContent(/About/g);
```
`/About/g`에서 `g`는 무엇을 의미하나요?

solmatches replied:
아 대소문자를 구분하지 않고 텍스트를 찾기위해 `/g` 를 사용하였었습니다만 잘못 사용하고 있었네요... 🥲
대소문자 구분을 위해서는 `/i`를 사용했었어야 했는데 왜인지 `/g`를 사용해야 한다고 인식하고 있었고 대소문자를 구분하지 않고 찾기를 의도했는데도 정규식에 들어간 문자는 대소문자를 구분한 값을 넣었네요 ㅠㅜㅜ 덕분에 잘못된 사실을 기억하고 있음을 깨달았습니다ㅠㅠ

hannut91 commented on src/components/Header.test.jsx:
```diff
@@ -0,0 +1,18 @@
+
+  fireEvent.click(getByRole('heading', { level: 1 }));
+
+  expect(window.location.pathname).toBe('/');
```
이건 어떤 것을 테스트하고 있는 건가요?

solmatches replied:
Header에서 Link를 사용하고 있다보니 눌렀을 때 홈으로 이동하는 것을 테스트하고 싶었는데.. 아무래도 Header의 역할과 책임에 벗어난게 맞는것 같네요😂 Header를 눌렀을 때 홈으로 이동하는 것을 테스트하고 싶다면 App에서 해주는게 맞는걸까요?

hannut91 replied:
`window.location.pathname`의 현재 위치는 `/`이니 변하지 않는 값인 것 같아요. 눌렀을 때 테스트는 다르게 테스트해야 할 것 같아요

hannut91 commented:
API호출을 하면서 만약에 에러가 발생한 경우도 고려해 보세요

solmatches committed:
대소문자를 구분하지 않는 정규식을 사용하도록 AboutPage 테스트 수정

- 문제: 대소문자를 구분하는 정규식으로 `/g`를 잘못 사용함
  - `/g`는 전체 문자에서 모든 일치를 반환하는 정규식.
  - 대소문자 구분을 위해서는 `/i`를 사용해야함.

toHaveTextContent 문서에도 나와있음
https://github.com/testing-library/jest-dom#tohavetextcontent

solmatches committed:
라이브러리와 로컬 파일의 import를 빈줄로 구분

solmatches committed:
Header 테스트에서 location.pathname 테스트 제거

Header 컴포넌트가 url 주소 영역까지 관리하는 것은 역할을 침범 하는 것 같아 제거

solmatches committed:
레스토랑 상세 페이지 제목으로 레스토랑 이름이 있음을 확인하는 테스트 작성

solmatches committed:
레스토랑 상세 페이지 제목에 음식점 이름 노출 구현

solmatches committed:
레스토랑 상세 페이지에 주소, 메뉴 노출 테스트 작성

solmatches committed:
레스토랑 상세 페이지 제목, 주소, 메뉴 리스트 구현

solmatches committed:
레스토랑 목록 클릭시 dispatch 호출 테스트 추가

solmatches committed:
레스토랑 목록을 누르면 상세 페이지로 이동하도록 Link 추가 구현

- RestaurantsContainer, RestaurantsPage 테스트업데이트
  - render될 엘리먼트에 MemoryRouter 추가

solmatches committed:
레스토랑 상세 페이지 Route 추가

solmatches committed:
레스토랑 상세 페이지 진입시 dispatch 호출 테스트 작성

solmatches committed:
index.html head에 base 태그 추가

/restaurant/${restaurantId}로 접근시
/restaurants/main.js 에서 main.js를 찾는 문제 발생
- index.html 에서 main.js를 절대 경로로 지정해놨기 때문
- `/main.js`로 변경해주거나 base 태그를 추가하여 문제 해결 가능

참고 사항
- base 태그는 모든 상대 URL에 적용될 기준이되어 전역으로 지정된다.
- 하나의 base 태그만 사용해야하고 닫는 태그를 넣으면 안됨 e.g) <base /> (X)

solmatches committed:
레스토랑 상세페이지 진입시 레스토랑 목록 불러오기 dispatch 호출 구현

- 해당 테스트에 컴포넌트를 렌더하지 않아 수정

solmatches committed:
레스토랑 상세 정보 불러오기 actions 성공, 실패 테스트 작성

solmatches committed:
레스토랑 상세 정보 불러오기 actions, api 구현

- mock api에도 fetchRestaurantDetails 추가

solmatches committed:
index.html에 적어두었던 base 태그 관련 주석 제거

solmatches committed:
레스토랑 상세 페이지 컨테이너 정보 렌더 테스트 작성

solmatches committed:
레스토랑 상세 페이지의 렌더부분을 레스토랑 상세 컨테이너로 이동

- 메뉴 렌더에 대한 테스트에서 오류 발생하여 수정

solmatches committed:
레스토랑 상세 컨테이너에 리덕스 데이터 연결 구현

- reduce에 `setRestaurantDetails`와 initialState 업데이트

- reducer 테스트 중 state가 없을 때 initialState 값도 업데이트

solmatches committed:
리듀서의 setRestaurantDetails 테스트 작성

solmatches committed:
리듀서 초기값을 공통으로 사용하도록 fixtures 로 이동

solmatches committed:
Provider와 BrowserRouter의 위치 변경

검색을 해보니 두 컴포넌트의 위치가 기능적으로 영향을 주는 것은 없지만 강의자료나 예제코드 등을 접하다보니
Provider가 가장 상위에 위치하는 경우가 일반적인 것 처럼 보여 변경함

solmatches committed:
레스토랑 상세 정보 통신시 pending, failure, success에 따른 화면 구현

hannut91 commented on src/RestaurantDetailsContainer.jsx:
```diff
@@ -0,0 +1,32 @@
+    return <div>loading...</div>;
+  }
+
+  if (failure) {
+    return <div>레스토랑 상세 정보를 가져오지 못했어요</div>;
+  }
```
좋습니다! 뭔가 잘못되었다는 것을 사용자가 알 수 있게 되었네요. 그런데 사용자는 이제 뭘 해야 될까요? 그냥 나가면 될까요?
그래서 무엇을 해야되는지도 알려주세요

hannut91 commented on src/RestaurantDetailsContainer.jsx:
```diff
@@ -0,0 +1,32 @@
+    return <div>레스토랑 상세 정보를 가져오지 못했어요</div>;
+  }
+
+  return (
+    <div>
+      <h2>{details?.name}</h2>
+      <p>
+        주소:
+        {' '}
+        {details?.address}
+      </p>
+      <h3>메뉴</h3>
+      <ul>
+        {details?.menuItems.map((menu) => (
+          <li key={menu.id}>{menu.name}</li>
+        ))}
+      </ul>
+    </div>
+  );
```
details가 없는 경우에는 아예 렌더하지 않도록 해서 `?`를 없애는게 좋을 것 같아요. 모든 코드에 `?`가 붙어서 이해하기 어려운 것 같습니다

hannut91 commented on src/RestaurantDetailsContainer.test.jsx:
```diff
@@ -0,0 +1,33 @@
+  it('레스토랑 이름인 페이지 제목이 보인다.', async () => {
+    const { findByRole } = render(<RestaurantDetailsContainer />);
+
+    expect(await findByRole('heading', { level: 2 })).toHaveTextContent(RESTAURANT_DETAILS.name);
```
`await`이 붙은 이유는 무엇인가요?

hannut91 reviewed (commented):
지역 목록와 카테고리 목록을 가져올 때, 두 데이터는 서로 연관이 없는데 지역 목록을 가져오고 나서야 카테고리 목록을 가져오고 있습니다. 동시에 가져와서 더 빠르게 처리할 수 있도록 수정해 주세요

solmatches committed:
통신 상태를 관리하는 status값이 추가되어 테스트 코드에 반영

solmatches committed:
Header에 onClick prop를 전달하여 라우팅이 이루어지도록 변경시도

- 테스트코드에 a 태그와 onClick에 대한 테스트 추가
